File: TidbitsTrivia/Core/Engine/Scoring.swift
import Foundation

/// Speed-aware scoring. Base points for a correct answer, a speed bonus
/// that rewards (but never requires) quickness, and a streak multiplier
/// that makes a hot run feel hot — capped so it stays a thrill, not a
/// runaway (engagement theory: variable-but-bounded reward).
enum Scoring {
    static let base = 100
    static let maxSpeedBonus = 100
    static let streakStep = 0.1     // +10% per consecutive correct
    static let maxStreakMultiplier = 2.0

    /// Points for one answer.
    /// - secondsTaken / budget normalizes speed across modes.
    static func points(correct: Bool, secondsTaken: Double, budget: Double, streak: Int) -> Int {
        guard correct else { return 0 }
        let speedFraction = max(0, min(1, 1 - secondsTaken / max(budget, 0.001)))
        let speed = Int(Double(maxSpeedBonus) * speedFraction)
        let mult = min(maxStreakMultiplier, 1 + Double(max(0, streak - 1)) * streakStep)
        return Int(Double(base + speed) * mult)
    }
}

File: windows/Tidbits.Core/Engine/Scoring.cs
namespace Tidbits.Core.Engine;

/// Speed-aware scoring (port of Core/Engine/Scoring.swift). Base points + a speed
/// bonus that rewards but never requires quickness + a bounded streak multiplier
/// (a hot run feels hot, capped so it stays a thrill not a runaway). Int casts
/// truncate toward zero, matching Swift's `Int(...)`.
public static class Scoring
{
    public const int Base = 100;
    public const int MaxSpeedBonus = 100;
    public const double StreakStep = 0.1;        // +10% per consecutive correct
    public const double MaxStreakMultiplier = 2.0;

    public static int Points(bool correct, double secondsTaken, double budget, int streak)
    {
        if (!correct) return 0;
        var speedFraction = Math.Max(0, Math.Min(1, 1 - secondsTaken / Math.Max(budget, 0.001)));
        var speed = (int)(MaxSpeedBonus * speedFraction);
        var mult = Math.Min(MaxStreakMultiplier, 1 + Math.Max(0, streak - 1) * StreakStep);
        return (int)((Base + speed) * mult);
    }
}
